[cluster-test] Make dockerfile for cluster test

andll opened the pull request:
This diff introduces single dockerfile for cluster test in docker/cluster-test/cluster-test.Dockerfile

Since we want to support incremental build, this docker file is generated via generate.sh script

After image is built it can be ran as this:
```
   docker run -t --rm cluster-test:latest <params>
```

For example

```
   docker run -t --rm cluster-test:latest --workplace cluster-test --healthcheck
```

andll commented:
Circle is failing because cluster-test dir was renamed, see demo in https://github.com/libra/libra/pull/1710

andll commented on docker/cluster-test/build.sh:
```diff
@@ -12,7 +12,7 @@ fi
 rm -f cluster_test_docker_builder_cluster_test
 
-docker build -f $DIR/cluster_test_builder.Dockerfile $DIR/../.. --tag libra_cluster_test_builder $PROXY
+docker build -f $DIR/cluster-test.builder.Dockerfile $DIR/../.. --tag libra_cluster_test_builder $PROXY
```
Yeah, we still need incremental build for container - many things in cluster test can only be tested in real environment

sausagee replied:
Should we rename this script to to reflect the build flow here?  This is to keep dir structure consistent with other images.  Other images (validator, mint...) have a single dockerfile and a build.sh script calls it directly.

sausagee commented on .circleci/config.yml:
```diff
@@ -116,6 +116,16 @@ jobs:
+      - run:
+          # generate dockerfile; diff config to existing one
+          name: Validate cluster-test dockerfiles
+          command: ./docker/cluster-test/generate.sh && git diff --exit-code
```
We can skip the `git diff --exit-code`.  The `build_teardown` step below has a check to screen changed files and will error out if the check fails.

andll commented on docker/cluster-test/cluster-test.builder.Dockerfile:
```diff
@@ -1,13 +1,12 @@
-# To use http/https proxy while building, use:
-# docker build --build-arg https_proxy=http://fwdproxy:8080 --build-arg http_proxy=http://fwdproxy:8080
+## To use http/https proxy while building, use:
+## docker build --build-arg https_proxy=http://fwdproxy:8080 --build-arg http_proxy=http://fwdproxy:8080
```
This is pretty standard (every docker file has them).
If we don't want those lets just remove from every file in separate commit

But imho, we can just keep it - maybe we can't build from devvm, but some other use case might want proxy

andll commented:
Addressed your comments
- Added explanation on why specific GCC version is needed
- Added comment on using preprocessor for dockerfile generation
- Removed `git diff --exit-code`

For naming the script - I would keep it as `build.sh`, because I actually want CI to test incremental build if something changes in docker files - for cluster test I actually routinely use so I want make sure CI keeps it working

andll commented on .dockerignore:
```diff
@@ -1,5 +1,4 @@
 *~
 **/.terraform/
-terraform/
```
So removing it from dockerignore won't just copy whole dir into container, you need to use `COPY` in dockerfile to choose what you want to copy
It just adds terraform dir to docker daemon context. If dir is super large it can be problem, but terraform dir is not that large...

sausagee replied:
In other images' build file, we do a full copy of the git repo.  This will cause their build to miss layer cache.

sausagee replied:
Looking at it 2nd time, this is fine because most of us don't touch terraform config, and it will be transparent.

sausagee commented:
@bors-libra delegate+

bors-libra commented:
:v: @andll can now approve this pull request

andll commented:
@bors-libra r=sausagee

bors-libra commented:
:pushpin: Commit 00c4cfe has been approved by `sausagee`

<!-- @bors-libra r=sausagee 00c4cfe5d5de7b9358330b71dd823ed2830cf30d -->

bors-libra commented:
:hourglass: Testing commit 00c4cfe5d5de7b9358330b71dd823ed2830cf30d with merge a6dd52d8103efe28ccb3e454d71d6dae111d0611...

andll commented:
@bors-libra retry

bors-libra commented:
:hourglass: Testing commit 00c4cfe5d5de7b9358330b71dd823ed2830cf30d with merge e2fb078a31b3d2a9f57db187b380050bbfbea20d...

bors-libra commented:
:sunny: Test successful - [checks-circle_commit_workflow](https://circleci.com/workflow-run/61b3f031-45a8-44ef-a7e9-8210f8d2479c?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link)
Approved by: sausagee
Pushing e2fb078a31b3d2a9f57db187b380050bbfbea20d to master...
